xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -209,160 +220,151 @@ bool CalculatesExpectation(const HloInstruction* instr) {
-                                     absl::flat_hash_set<int>& visited_instrs,
-                                     int operand_idx,
-                                     HloInstruction* transpose) {
-  visited_instrs.emplace(instr->unique_id());
+// Returns whether target can be reached from instr by recursively traversing
+// the graph across converts, bitcasts and reshapes.
+bool FindTargetRecursive(
+    const HloInstruction* instr, const HloInstruction* target,
+    absl::flat_hash_set<const HloInstruction*>& visited_instrs,
+    const HloInstruction* transpose) {
+  visited_instrs.emplace(instr);
   const absl::flat_hash_set<HloOpcode> supported_ops = {
       HloOpcode::kConvert, HloOpcode::kBitcast, HloOpcode::kReshape};
-  // Look for the Custom Call among the users of instr.
+  if (instr == target) {
+    return true;
+  }
+  // Look for target among the users of instr.
   for (HloInstruction* user : instr->users()) {
-    if (user == custom_call && user->operand(operand_idx) == instr) {
-      return instr;
-    }
     if ((supported_ops.contains(user->opcode()) || user == transpose) &&
-        !visited_instrs.contains(user->unique_id())) {
-      return FindOperandRecursive(user, custom_call, visited_instrs,
-                                  operand_idx, transpose);
+        !visited_instrs.contains(user)) {
+      return FindTargetRecursive(user, target, visited_instrs, transpose);
     }
   }
-  // Ascend the graph if the Custom Call is not found and instr is a
-  // conversion, bitcast or reshape.
+  // Ascend the graph if target is not found and instr is a conversion, bitcast
+  // or reshape.
   if (supported_ops.contains(instr->opcode())) {
-    return FindOperandRecursive(instr->mutable_operand(0), custom_call,
-                                visited_instrs, operand_idx, transpose);
+    return FindTargetRecursive(instr->operand(0), target, visited_instrs,
+                               transpose);
   }
-  return nullptr;
+  return false;
 }
-HloInstruction* FindOperandRecursive(HloInstruction* instr,
-                                     HloInstruction* custom_call,
-                                     int operand_idx,
-                                     const NormMetaDataMap& norm_meta_data) {
-  absl::flat_hash_set<int> visited_instrs;
-  auto custom_call_meta_data = norm_meta_data.find(custom_call->unique_id());
-  if (custom_call_meta_data == norm_meta_data.end()) {
-    return nullptr;
-  }
-  return FindOperandRecursive(instr, custom_call, visited_instrs, operand_idx,
-                              custom_call_meta_data->second.x_transpose);
+
+bool FindTargetRecursive(const HloInstruction* custom_call,
+                         const HloInstruction* instr,
+                         const HloInstruction* target) {
+  absl::flat_hash_set<const HloInstruction*> visited_instrs;
+  return FindTargetRecursive(instr, target, visited_instrs,
+                             custom_call->operand(0)->operand(0));
```

Comment:
`custom_call->operand(0)->operand(0)` isn't necessarily the transpose, because it's possible no transpose was added.

I think the best way to fix this would be to have NormMetadata store pointers to the pre-tranposed input and the scale. Then, the caller can just pass those in as the target. Since the input is pre-transpose, you don't need to check the transpose at all in `FindTargetRecursive`.

Also perhaps this overload of `FindTargetRecursive` should be called `FindTarget`, and keep the other called `FindTargetRecursive`.